ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs

```diff
@@ -273,6 +289,46 @@ instance ShelleyCompatible proto era => QueryLedger (ShelleyBlock proto era) whe
                 , SL._retiring = Map.restrictKeys (SL._retiring dpsPState) poolIds
                 }
             Nothing -> dpsPState
+        GetStakeSnapshots mPoolIds ->
+          let SL.SnapShots
+                { SL._pstakeMark
+                , SL._pstakeSet
+                , SL._pstakeGo
+                } = SL.esSnapshots . SL.nesEs $ st
+
+              -- | Sum all the stake that is held by the pool
+              getPoolStake :: KeyHash 'StakePool crypto -> SL.SnapShot crypto -> SL.Coin
+              getPoolStake hash ss = VMap.foldMap fromCompact s
+                where
+                  SL.Stake s = SL.poolStake hash (SL._delegations ss) (SL._stake ss)
+
+              getPoolStakes :: Set (KeyHash 'StakePool crypto) -> Map (KeyHash 'StakePool crypto) (StakeSnapshot crypto)
+              getPoolStakes poolIds = Map.fromSet mkStakeSnapshot poolIds
+                where mkStakeSnapshot poolId = StakeSnapshot
+                        { ssMarkPool   = getPoolStake poolId _pstakeMark
+                        , ssSetPool    = getPoolStake poolId _pstakeSet 
+                        , ssGoPool     = getPoolStake poolId _pstakeGo  
+                        }
+
+              getAllStake :: SL.SnapShot crypto -> SL.Coin
+              getAllStake (SL.SnapShot stake _ _) = VMap.foldMap fromCompact (SL.unStake stake)
+          in
+
+          case mPoolIds of
+            Nothing ->
+              StakeSnapshots
+                { ssStakeSnapshots = getPoolStakes (Set.fromList (VMap.elems (SL._delegations _pstakeMark)))
```

Comment:
I'm assuming that if I get the `elems` of `_delegations _pstakeMark` that it would give me all of the pool ids.

I'm not confident that this is true.  Please validate.

Reply:
That is correct @newhoggy. `_delegations _pstakeMark` is a map from stake credentials to stake pool IDs. You can tell from they type, if you happen to know that the stake pool IDs are the hash of the stake pool operator's cold verification key:

https://github.com/input-output-hk/cardano-ledger/blob/93b5d5421a5b5c3a3b6ad89a74e8bdb2ccc9258c/eras/shelley/impl/src/Cardano/Ledger/Shelley/EpochBoundary.hs#L157

Reply:
Thanks!